### PR TITLE
Python: Fix 'unused import' for doctests and typehints.

### PR DESCRIPTION
**Pull request by markshannon**
Don't give alerts for 'unused import' when the imported module is used in either a doctest or a typehint.

### CODE REVIEW

File: change-notes/1.20/analysis-python.md [hunk 1]
```diff
@@ -16,6 +16,7 @@
 
  | **Query**                  | **Expected impact**    | **Change**                                                       |
 |----------------------------|------------------------|------------------------------------------------------------------|
+| Unused import (`py/unused-import`) | Fewer false positive results | Results where the imported module is used in a doctest string are no longer reported |
```
Comment (felicity-semmle) [inline change-notes/1.20/analysis-python.md]: Thanks for the change note. Please can we have "doctest" in backticks and a period at the end of the sentence.

File: python/ql/src/Imports/UnusedImport.ql [hunk 2]
```diff
@@ -41,6 +41,37 @@ predicate all_not_understood(Module m) {
     )
 }
 
+predicate imported_module_used_in_doctest(Import imp) {
+    exists(string modname |
+        ((Name)imp.getAName().getAsname()).getId() = modname
+        and
+        /* Look for doctests containing the patterns:
+         * >>> …name…
```
Comment (taus-semmle) [inline python/ql/src/Imports/UnusedImport.ql]: Unicode in comments! Very advanced. :)

File: python/ql/src/Imports/UnusedImport.ql [hunk 3]
```diff
@@ -41,6 +41,37 @@ predicate all_not_understood(Module m) {
...
+predicate imported_module_used_in_typehint(Import imp) {
+    exists(string modname |
+        ((Name)imp.getAName().getAsname()).getId() = modname
+        and
+        /* Look for typehints containing the patterns:
+         * # type: …name…
+         */
+        exists(Comment tyephint |
+            tyephint.getLocation().getFile() = imp.getScope().(Module).getFile() and
+            tyephint.getText().regexpMatch("# type:.*" + modname + ".*")
```
Comment (taus-semmle) [inline python/ql/src/Imports/UnusedImport.ql]: `tyephint` looks like a typo to me. (But it's consistent, I'll admit.)

File: python/ql/src/Imports/UnusedImport.ql [hunk 4]
```diff
@@ -41,6 +41,37 @@ predicate all_not_understood(Module m) {
...
+            doc.getEnclosingModule() = imp.getScope() and
+            doc.isDocString() and
+            doc.getText().regexpMatch("[\\s\\S]*(>>>|\\.\\.\\.).*" + modname + "[\\s\\S]*")
+        )
+    )
+}
+
+predicate imported_module_used_in_typehint(Import imp) {
+    exists(string modname |
+        ((Name)imp.getAName().getAsname()).getId() = modname
```
Comment (taus-semmle) [inline python/ql/src/Imports/UnusedImport.ql]: Any reason for not using an infix cast? (Also, the `.getAName().getAsname().(Name)` invocation always makes me pause. Surely we could make this a bit less redundant?)

**Pull request merged**